### PR TITLE
Versioning System

### PR DESCRIPTION
**Pull request by tithanayut**
## Why did you create this PR

In order to build a prosperous and sustainable future, a solid foundation is needed upon which versioning can be realized. 

This PR implements versioning system for cugetreg-web, cugetreg-api and future apps/libs.

## What did you do

- Will be discussed in meeting

## Required Configurations
- Generate a key pair, add public key to repo's Deploy keys and private key to Action secrets name `VERSIONING_DEPLOY_KEY` **with write access**. A key pair comment must be `git@github.com:thinc-org/cugetreg.git` ([see this](https://github.com/webfactory/ssh-agent#support-for-github-deploy-keys)).
- (optional) Install [changeset-bot](https://github.com/apps/changeset-bot) app

## Demo

[https://github.com/tithanayut/cugetreg-playground](https://github.com/tithanayut/cugetreg-playground)

## Test Cases

- [ ] Command `pnpm changeset` prompts changeset dialog
- [ ] Creating a PR without changeset triggers [changeset-bot](https://github.com/apps/changeset-bot) to comment on the PR ([example](https://github.com/tithanayut/cugetreg-playground/pull/1))
- [ ] Creating a PR with changeset triggers versioning workflow to create Version Packages PR ([example](https://github.com/tithanayut/cugetreg-playground/pull/9))
- [ ] Creating a new PR with another changeset triggers versioning workflow to update changelog in previously created Version Packages PR. ([example](https://github.com/tithanayut/cugetreg-playground/pull/7))
- [ ] Version Packages PR created correctly deletes all changesets, bumps version number, and updates CHANGELOG file(s) of relevant apps/libs ([example](https://github.com/tithanayut/cugetreg-playground/pull/9/files))  
- [ ] Merging Version Packages PR results in the creation of git tag in the format of `app-name@version` of relevant apps/libs ([example](https://github.com/tithanayut/cugetreg-playground/tags)) (`versioning` workflow)
- [ ] Merging Version Packages PR results in the publish of docker image in the format of `app-name:version` of relevant apps ([example](https://github.com/tithanayut/cugetreg-playground/pkgs/container/cugetreg-web)) (`publish` workflow)
- [ ] Merging Version Packages PR results in the creation of GitHub release with correct release note and CHANGELOG file ([example](https://github.com/tithanayut/cugetreg-playground/releases)) (`publish` workflow)

## Checklist

- [x] Deploy a demo
- [ ] Check browsers compatibility (Not applicable)
- [ ] Wrote coverage tests (Not applicable)

## Related links

- https://pnpm.io/using-changesets
- https://github.com/changesets/bot
- https://github.com/changesets/action

### CODE REVIEW

**Comment by tithanayut** [comment]
I think this would not work (for now) due to changes in #441. Also, I think we should discuss about the deployment flow on GitOps to further tune this as well.

File: .github/workflows/versioning.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        id: changesets
+        uses: changesets/action@v1.3.0
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Setup Deploy Key
+        uses: webfactory/ssh-agent@v0.5.4
+        if: steps.changesets.outputs.hasChangesets == 'false'
+        with:
+          ssh-private-key: ${{ secrets.VERSIONING_DEPLOY_KEY }}
```
Comment (saenyakorn) [inline .github/workflows/versioning.yaml:55]: Who own this ssh private key?

File: .github/workflows/versioning.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                  | head -1 \
+                  | awk -F: '{ print $2 }' \
+                  | sed 's/[", ]//g')
+            TAG=$APP@$PACKAGE_VERSION
+
+            if ! git rev-parse "$TAG" > /dev/null 2>&1; then
+              git tag $APP@$PACKAGE_VERSION
+              git push origin $APP@$PACKAGE_VERSION
+            fi 
+          done
```
Comment (saenyakorn) [inline .github/workflows/versioning.yaml:73]: I think you could write a bash file for these commands.

Comment (saenyakorn) [reply .github/workflows/versioning.yaml:73]: Also, comment line by line what is the result or expectation from that line.

File: .github/workflows/publish.yaml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      - name: Build Docker images
+        run: npx nx run ${{ steps.meta.outputs.APP }}:docker-build
+
+      - name: Tag and push Docker images
+        run: |
+          docker image tag ${{ steps.meta.outputs.APP }}:latest ghcr.io/thinc-org/${{ steps.meta.outputs.APP }}:${{ steps.meta.outputs.VERSION }}
+          docker push ghcr.io/thinc-org/${{ steps.meta.outputs.APP }}:${{ steps.meta.outputs.VERSION }}
+
+          docker image tag ${{ steps.meta.outputs.APP }}:latest ghcr.io/thinc-org/${{ steps.meta.outputs.APP }}:latest
+          docker push ghcr.io/thinc-org/${{ steps.meta.outputs.APP }}:latest
```
Comment (saenyakorn) [inline .github/workflows/publish.yaml:76]: Bash file for these command

File: .github/workflows/versioning.yaml (hunk 2 above)
Comment (bombnp) [reply .github/workflows/versioning.yaml:73]: u could put the file in `tools/changesets/**`or something

File: .github/workflows/publish.yaml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+      - name: Docker login
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ github.token }}
+
+      - name: Build Docker images
+        run: npx nx run ${{ steps.meta.outputs.APP }}:docker-build
```
Comment (saenyakorn) [inline .github/workflows/publish.yaml:68]: use [run-many](https://nx.dev/packages/nx/documents/run-many) command instead

```
nx run-many --parallel --target=serve --projects=${{ steps.meta.outputs.APP }}
```

Note: `APP` should follow by this format, `app1,app2,app3`

File: .github/workflows/versioning.yaml (hunk 2 above)
Comment (saenyakorn) [reply .github/workflows/versioning.yaml:73]: I suggest writing this thing as a programming language like `python` or `typescript` for more readable code and more maintainability. 

This line is tough to read at first sight, especially for beginners. 
```
PACKAGE_VERSION=$(cat ./apps/$APP/package.json \
                  | grep version \
                  | head -1 \
                  | awk -F: '{ print $2 }' \
                  | sed 's/[", ]//g')
            TAG=$APP@$PACKAGE_VERSION
```

File: .github/workflows/versioning.yaml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Setup Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: 16.17.0
```
Comment (saenyakorn) [inline .github/workflows/versioning.yaml:22]: i suggest using `node-version` from matrix. For example

```yaml
jobs:
  strategy:
        matrix:
          node-version: [16.17.0]
  steps:
    // ...
    - name: Setup Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v3
        with:
          node-version: ${{ matrix.node-version }}
```

**Comment by tithanayut** [comment]
#456

File: .github/workflows/publish.yaml (hunk 4 above)
Comment (tithanayut) [reply .github/workflows/publish.yaml:68]: This workflow will only run for one app at a time.

File: .github/workflows/publish.yaml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+          docker image tag ${{ steps.meta.outputs.APP }}:latest ghcr.io/thinc-org/${{ steps.meta.outputs.APP }}:latest
+          docker push ghcr.io/thinc-org/${{ steps.meta.outputs.APP }}:latest
+
+      - name: Generate release note
+        run: |
+          git diff --color-words --unified=0 HEAD HEAD~ apps/${{ steps.extract.outputs.APP }}/CHANGELOG.md \
+            | tail -n +8 \
+            | sed -r "s/\x1B\[([0-9]{1,3}(;[0-9]{1,2})?)?[mGK]//g" > RELEASE.md
+
```
[the remark this answers is not in the document]
Comment (tithanayut) [reply .github/workflows/publish.yaml:83]: Should use Python or JS instead.

File: .github/workflows/versioning.yaml (hunk 1 above)
Comment (tithanayut) [reply .github/workflows/versioning.yaml:55]: We could generate a new key pair and add it to the repo deploy key for this as mentioned in `Required Configurations` in the PR description.

**Pull request closed without merging**